tests/optim/test_oss.py

```diff
@@ -662,3 +662,74 @@ def test_state_dict_distributed():
     )
+
+
+def run_ddp_parity(rank, world_size, backend, temp_file_name):
+    url = "file://" + temp_file_name
+    dist.init_process_group(init_method=url, backend=backend, rank=rank, world_size=world_size)
+
+    device = torch.device("cuda")
+    torch.cuda.set_device(rank)
+    torch.manual_seed(rank)
+    np.random.seed(rank)
+
+    def check_optimizer_equivalence(optimizer: Type[torch.optim.Optimizer]):
+        # Any model works. Add one different buffer per rank
+        model = torch.nn.Sequential(torch.nn.Linear(2, 3), torch.nn.Linear(3, 3), torch.nn.Linear(3, 3),)
+        model.register_buffer("test_buffer", torch.ones((1)) * rank)
+        model.to(device)
+
+        sharded_optimizer = optim.OSS(params=model.parameters(), optim=optimizer, lr=1e-3)
+        sharded_ddp_model = DDP(module=model, device_ids=[rank], broadcast_buffers=True)
+
+        ddp_model_single = copy.deepcopy(model)
+        ddp_optimizer = optimizer(ddp_model_single.parameters(), lr=1e-3)
+        ddp_model = DDP(ddp_model_single, device_ids=[rank], broadcast_buffers=True)
+
+        def check_same_model_params():
+            for pg, ddp_pg in zip(sharded_optimizer.param_groups, ddp_optimizer.param_groups):
+                for p, ddp_p in zip(pg["params"], ddp_pg["params"]):
+                    assert torch.allclose(
+                        p, ddp_p, atol=1e-3
+                    ), f"Model parameters differ in between Pytorch optim and OSS {p} {ddp_p}"
+
+            for b, ddp_b in zip(sharded_ddp_model.buffers(), ddp_model.buffers()):
+                assert torch.allclose(b, ddp_b, atol=1e-3), "Model buffers differ in between Pytorch optim and OSS"
+
+        # The model should be synchronized in between the ranks at construction time, check that
+        check_same_model_params()
+
+        # The models should stay the same in between the ranks
+        for i in range(20):
+            input_tensor = torch.rand((64, 2)).to(device)
+
+            def closure_ddp(input_tensor=input_tensor):
+                ddp_optimizer.zero_grad()
+                ddp_loss = ddp_model(input_tensor).abs().sum()
+                ddp_loss.backward()
+                return ddp_loss
+
+            def closure_sharded(input_tensor=input_tensor):
+                sharded_optimizer.zero_grad()
+                sharded_loss = sharded_ddp_model(input_tensor).abs().sum()
+                sharded_loss.backward()
+                return sharded_loss
+
+            _ = ddp_optimizer.step(closure=closure_ddp)
+            _ = sharded_optimizer.step(closure=closure_sharded)
+
+            check_same_model_params()
+
+    for opt in [torch.optim.SGD, torch.optim.Adam]:
+        check_optimizer_equivalence(opt)
+
+    dist.destroy_process_group()
+
+
+@skip_if_no_cuda
+@skip_if_single_gpu
+def test_ddp_parity():
+    temp_file_name = tempfile.mkstemp()[1]
+    world_size = torch.cuda.device_count()
```

Comment:
need to log the world size in case it fails and we know what was the value in CI?